In [1]:
import os

for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/minuteCaloriesNarrow_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/weightLogInfo_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/minuteMETsNarrow_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/hourlyCalories_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/heartrate_seconds_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/minuteSleep_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/hourlyIntensities_merged.csv
/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitaba

# Bellabeat Case Study
## How Can a Wellness Company Play It Smart?

### Data Analytics Portfolio Project

**Tools:** Python (Pandas), Microsoft Excel

### Project Overview
This project analyzes Fitbit smart-device data to identify trends in user activity and sleep behavior and explores how these insights could help Bellabeat improve its marketing strategy and user engagement.

### Business Questions
1. What are some trends in smart-device usage?
2. How could these trends apply to Bellabeat customers?
3. How could these trends help influence Bellabeat's marketing strategy?

## 1. Data Preparation

In this section, the Fitbit activity and sleep datasets are loaded and inspected to understand their structure, data types, completeness, and overall quality before cleaning.

In [1]:
import pandas as pd

daily_activity_path = "/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/dailyActivity_merged.csv"

daily_activity = pd.read_csv(daily_activity_path)

daily_activity.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,3/25/2016,11004,7.11,7.11,0.0,2.57,0.46,4.07,0.0,33,12,205,804,1819
1,1503960366,3/26/2016,17609,11.55,11.55,0.0,6.92,0.73,3.91,0.0,89,17,274,588,2154
2,1503960366,3/27/2016,12736,8.53,8.53,0.0,4.66,0.16,3.71,0.0,56,5,268,605,1944
3,1503960366,3/28/2016,13231,8.93,8.93,0.0,3.19,0.79,4.95,0.0,39,20,224,1080,1932
4,1503960366,3/29/2016,12041,7.85,7.85,0.0,2.16,1.09,4.61,0.0,28,28,243,763,1886


## 2. Data Cleaning

The activity datasets were combined and checked for duplicate user-date records, missing values, and date consistency.

The two activity periods overlapped on April 12, 2016, resulting in 24 duplicate user-date records. The records from the later collection period were retained, resulting in a cleaned activity dataset of 1,373 records from 35 users.

The sleep dataset was also checked and cleaned by removing 3 exact duplicate records, leaving 410 sleep records from 24 users.

Zero-step activity records were retained because they may represent genuine inactivity or incomplete device usage rather than definite data errors.

In [2]:
print("Rows and columns:")
print(daily_activity.shape)

print("\nData types:")
print(daily_activity.dtypes)

print("\nMissing values:")
print(daily_activity.isnull().sum())

print("\nDuplicate rows:")
print(daily_activity.duplicated().sum())

Rows and columns:
(457, 15)

Data types:
Id                            int64
ActivityDate                 object
TotalSteps                    int64
TotalDistance               float64
TrackerDistance             float64
LoggedActivitiesDistance    float64
VeryActiveDistance          float64
ModeratelyActiveDistance    float64
LightActiveDistance         float64
SedentaryActiveDistance     float64
VeryActiveMinutes             int64
FairlyActiveMinutes           int64
LightlyActiveMinutes          int64
SedentaryMinutes              int64
Calories                      int64
dtype: object

Missing values:
Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        

In [3]:
daily_activity["ActivityDate"] = pd.to_datetime(
    daily_activity["ActivityDate"]
)

daily_activity.dtypes

Id                                   int64
ActivityDate                datetime64[ns]
TotalSteps                           int64
TotalDistance                      float64
TrackerDistance                    float64
LoggedActivitiesDistance           float64
VeryActiveDistance                 float64
ModeratelyActiveDistance           float64
LightActiveDistance                float64
SedentaryActiveDistance            float64
VeryActiveMinutes                    int64
FairlyActiveMinutes                  int64
LightlyActiveMinutes                 int64
SedentaryMinutes                     int64
Calories                             int64
dtype: object

In [4]:
print("Date range:")
print(daily_activity["ActivityDate"].min())
print(daily_activity["ActivityDate"].max())

print("\nNumber of unique users:")
print(daily_activity["Id"].nunique())

Date range:
2016-03-12 00:00:00
2016-04-12 00:00:00

Number of unique users:
35


In [5]:
second_activity_path = "/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_4.12.16-5.12.16/Fitabase Data 4.12.16-5.12.16/dailyActivity_merged.csv"

daily_activity_2 = pd.read_csv(second_activity_path)

daily_activity_2.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,4/12/2016,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,4/13/2016,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,4/14/2016,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,4/15/2016,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,4/16/2016,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [6]:
print("Rows and columns:")
print(daily_activity_2.shape)

print("\nData types:")
print(daily_activity_2.dtypes)

print("\nMissing values:")
print(daily_activity_2.isnull().sum())

print("\nDuplicate rows:")
print(daily_activity_2.duplicated().sum())

print("\nDate range:")
print(daily_activity_2["ActivityDate"].min(), "to", daily_activity_2["ActivityDate"].max())

print("\nNumber of unique users:")
print(daily_activity_2["Id"].nunique())

Rows and columns:
(940, 15)

Data types:
Id                            int64
ActivityDate                 object
TotalSteps                    int64
TotalDistance               float64
TrackerDistance             float64
LoggedActivitiesDistance    float64
VeryActiveDistance          float64
ModeratelyActiveDistance    float64
LightActiveDistance         float64
SedentaryActiveDistance     float64
VeryActiveMinutes             int64
FairlyActiveMinutes           int64
LightlyActiveMinutes          int64
SedentaryMinutes              int64
Calories                      int64
dtype: object

Missing values:
Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        

In [7]:
daily_activity_2["ActivityDate"] = pd.to_datetime(
    daily_activity_2["ActivityDate"]
)

daily_activity_2.dtypes

Id                                   int64
ActivityDate                datetime64[ns]
TotalSteps                           int64
TotalDistance                      float64
TrackerDistance                    float64
LoggedActivitiesDistance           float64
VeryActiveDistance                 float64
ModeratelyActiveDistance           float64
LightActiveDistance                float64
SedentaryActiveDistance            float64
VeryActiveMinutes                    int64
FairlyActiveMinutes                  int64
LightlyActiveMinutes                 int64
SedentaryMinutes                     int64
Calories                             int64
dtype: object

In [8]:
daily_activity = pd.concat(
    [daily_activity, daily_activity_2],
    ignore_index=True
)

print("Rows and columns:")
print(daily_activity.shape)

print("\nDate range:")
print(daily_activity["ActivityDate"].min(), "to", daily_activity["ActivityDate"].max())

print("\nUnique users:")
print(daily_activity["Id"].nunique())

Rows and columns:
(1397, 15)

Date range:
2016-03-12 00:00:00 to 2016-05-12 00:00:00

Unique users:
35


In [9]:
duplicate_user_dates = daily_activity.duplicated(
    subset=["Id", "ActivityDate"]
).sum()

print("Duplicate user-date records:", duplicate_user_dates)

Duplicate user-date records: 24


In [10]:
duplicate_rows = daily_activity[
    daily_activity.duplicated(
        subset=["Id", "ActivityDate"],
        keep=False
    )
].sort_values(["Id", "ActivityDate"])

duplicate_rows

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
18,1503960366,2016-04-12,224,0.14,0.14,0.000000,0.00,0.00,0.13,0.0,0,0,9,32,50
457,1503960366,2016-04-12,13162,8.50,8.50,0.000000,1.88,0.55,6.06,0.0,25,13,328,728,1985
37,1624580081,2016-04-12,6627,4.31,4.31,0.000000,0.00,0.00,4.31,0.0,0,0,89,604,706
488,1624580081,2016-04-12,8163,5.31,5.31,0.000000,0.00,0.00,5.31,0.0,0,0,146,1294,1432
59,1844505072,2016-04-12,0,0.00,0.00,0.000000,0.00,0.00,0.00,0.0,0,0,0,426,399
549,1844505072,2016-04-12,6697,4.43,4.43,0.000000,0.00,0.00,4.43,0.0,0,0,339,1101,2030
71,1927972279,2016-04-12,24,0.02,0.02,0.000000,0.00,0.00,0.02,0.0,0,0,3,161,942
580,1927972279,2016-04-12,678,0.47,0.47,0.000000,0.00,0.00,0.47,0.0,0,0,55,734,2220
83,2022484408,2016-04-12,6717,4.72,4.72,0.000000,3.23,0.22,1.27,0.0,41,5,67,551,1140
611,2022484408,2016-04-12,11875,8.34,8.34,0.000000,3.31,0.77,4.26,0.0,42,14,227,1157,2390


In [11]:
print("Period 1 - April 12 rows:", 
      (daily_activity["ActivityDate"] == "2016-04-12").sum())

print("Period 2 - April 12 rows:", 
      (daily_activity_2["ActivityDate"] == "2016-04-12").sum())

Period 1 - April 12 rows: 57
Period 2 - April 12 rows: 33


In [12]:
daily_activity = daily_activity.drop_duplicates(
    subset=["Id", "ActivityDate"],
    keep="last"
).reset_index(drop=True)

print("Rows and columns:")
print(daily_activity.shape)

print("\nDuplicate user-date records:")
print(daily_activity.duplicated(
    subset=["Id", "ActivityDate"]
).sum())

Rows and columns:
(1373, 15)

Duplicate user-date records:
0


In [13]:
print("Date range:")
print(daily_activity["ActivityDate"].min())
print(daily_activity["ActivityDate"].max())

print("\nRows after May 9:")
print(daily_activity[daily_activity["ActivityDate"] > "2016-05-09"])

Date range:
2016-03-12 00:00:00
2016-05-12 00:00:00

Rows after May 9:
              Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
461   1503960366   2016-05-10       12207       7.770000         7.770000   
462   1503960366   2016-05-11       12770       8.130000         8.130000   
463   1503960366   2016-05-12           0       0.000000         0.000000   
492   1624580081   2016-05-10        2969       1.930000         1.930000   
493   1624580081   2016-05-11        3134       2.040000         2.040000   
...          ...          ...         ...            ...              ...   
1312  8583815059   2016-05-12           0       0.000000         0.000000   
1341  8792009665   2016-05-10           0       0.000000         0.000000   
1370  8877689391   2016-05-10       10733       8.150000         8.150000   
1371  8877689391   2016-05-11       21420      19.559999        19.559999   
1372  8877689391   2016-05-12        8064       6.120000         6.120000   

    

In [14]:
print(daily_activity.isnull().sum())

Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64


In [1]:
daily_activity.describe()

NameError: name 'daily_activity' is not defined

In [2]:
import pandas as pd

# Load both activity files
daily_activity_1 = pd.read_csv(
    "/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_3.12.16-4.11.16/Fitabase Data 3.12.16-4.11.16/dailyActivity_merged.csv"
)

daily_activity_2 = pd.read_csv(
    "/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_4.12.16-5.12.16/Fitabase Data 4.12.16-5.12.16/dailyActivity_merged.csv"
)

# Convert dates
daily_activity_1["ActivityDate"] = pd.to_datetime(
    daily_activity_1["ActivityDate"]
)

daily_activity_2["ActivityDate"] = pd.to_datetime(
    daily_activity_2["ActivityDate"]
)

# Combine both periods
daily_activity = pd.concat(
    [daily_activity_1, daily_activity_2],
    ignore_index=True
)

# Remove overlapping user-date records
daily_activity = daily_activity.drop_duplicates(
    subset=["Id", "ActivityDate"],
    keep="last"
).reset_index(drop=True)

print("Final dataset shape:", daily_activity.shape)
print("Duplicate user-date records:",
      daily_activity.duplicated(
          subset=["Id", "ActivityDate"]
      ).sum())

Final dataset shape: (1373, 15)
Duplicate user-date records: 0


In [3]:
daily_activity.describe()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
count,1.373000e+03,1373,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000,1373.000000
mean,4.782326e+09,2016-04-19 04:24:17.829570304,7377.375091,5.289031,5.261930,0.128591,1.413146,0.546052,3.238128,0.001733,19.871085,13.603787,188.096140,1001.346686,2294.812090
min,1.503960e+09,2016-03-12 00:00:00,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2.320127e+09,2016-04-08 00:00:00,3321.000000,2.280000,2.270000,0.000000,0.000000,0.000000,1.730000,0.000000,0.000000,0.000000,117.000000,734.000000,1820.000000
50%,4.445115e+09,2016-04-19 00:00:00,7142.000000,5.030000,5.020000,0.000000,0.120000,0.200000,3.280000,0.000000,2.000000,6.000000,196.000000,1062.000000,2129.000000
75%,6.962181e+09,2016-04-30 00:00:00,10645.000000,7.570000,7.570000,0.000000,1.860000,0.780000,4.710000,0.000000,30.000000,18.000000,263.000000,1246.000000,2781.000000
max,8.877689e+09,2016-05-12 00:00:00,36019.000000,28.030001,28.030001,6.727057,21.920000,6.480000,12.510000,0.110000,210.000000,660.000000,720.000000,1440.000000,4900.000000
std,2.381544e+09,NaN,5198.133424,3.987341,3.973439,0.695630,2.622729,0.872581,2.104300,0.007800,31.817631,26.581555,113.095058,304.168197,725.526989


In [4]:
daily_activity["Id"].nunique()

35

In [5]:
user_days = daily_activity.groupby("Id")["ActivityDate"].count()

print(user_days.describe())

count    35.000000
mean     39.228571
std       9.490198
min       8.000000
25%      38.500000
50%      42.000000
75%      42.000000
max      62.000000
Name: ActivityDate, dtype: float64


## 3. Data Transformation and Analysis

After cleaning, the datasets were transformed to support the main business questions.

A DayOfWeek variable was created to examine weekly activity patterns. Average daily steps were calculated for each user, and users were grouped into project-defined activity levels:

- Low Activity: fewer than 5,000 average daily steps
- Moderate Activity: 5,000–9,999 average daily steps
- High Activity: 10,000 or more average daily steps

Additional analysis examined activity intensity, the relationship between daily steps and calories burned, and sleep duration across the week.

The activity and sleep datasets were also merged by user ID and date to explore relationships between activity and sleep behavior.

In [6]:
daily_activity["DayOfWeek"] = daily_activity["ActivityDate"].dt.day_name()

daily_activity[["ActivityDate", "DayOfWeek"]].head()

,ActivityDate,DayOfWeek
0,2016-03-25,Friday
1,2016-03-26,Saturday
2,2016-03-27,Sunday
3,2016-03-28,Monday
4,2016-03-29,Tuesday


In [7]:
steps_by_day = daily_activity.groupby("DayOfWeek")["TotalSteps"].mean()

steps_by_day

DayOfWeek
Friday       7187.532663
Monday       7541.319149
Saturday     7752.271357
Sunday       6606.725389
Thursday     7268.297436
Tuesday      7718.960199
Wednesday    7547.575758
Name: TotalSteps, dtype: float64

In [8]:
sedentary_by_day = daily_activity.groupby("DayOfWeek")["SedentaryMinutes"].mean()

sedentary_by_day

DayOfWeek
Friday       1020.266332
Monday       1029.718085
Saturday      968.934673
Sunday        999.730570
Thursday      984.902564
Tuesday      1012.208955
Wednesday     994.712121
Name: SedentaryMinutes, dtype: float64

In [9]:
zero_step_days = (daily_activity["TotalSteps"] == 0).sum()

print("Zero-step records:", zero_step_days)
print("Percentage of records:", 
      round(zero_step_days / len(daily_activity) * 100, 2), "%")

Zero-step records: 133
Percentage of records: 9.69 %


In [10]:
daily_activity["ActiveDay"] = daily_activity["TotalSteps"] > 0

daily_activity["ActiveDay"].value_counts()

ActiveDay
True     1240
False     133
Name: count, dtype: int64

In [11]:
active_days = daily_activity[daily_activity["ActiveDay"] == True]

active_days[
    [
        "VeryActiveMinutes",
        "FairlyActiveMinutes",
        "LightlyActiveMinutes",
        "SedentaryMinutes"
    ]
].mean()

VeryActiveMinutes        21.933065
FairlyActiveMinutes      14.530645
LightlyActiveMinutes    208.270968
SedentaryMinutes        960.793548
dtype: float64

In [12]:
daily_activity[["TotalSteps", "Calories"]].corr()

,TotalSteps,Calories
TotalSteps,1.000000,0.580077
Calories,0.580077,1.000000


In [13]:
activity_intensity = daily_activity[
    [
        "VeryActiveMinutes",
        "FairlyActiveMinutes",
        "LightlyActiveMinutes"
    ]
].mean()

activity_intensity

VeryActiveMinutes        19.871085
FairlyActiveMinutes      13.603787
LightlyActiveMinutes    188.096140
dtype: float64

In [14]:
sleep_day = pd.read_csv(
    "/kaggle/input/datasets/arashnic/fitbit/mturkfitbit_export_4.12.16-5.12.16/Fitabase Data 4.12.16-5.12.16/sleepDay_merged.csv"
)

print("Shape:", sleep_day.shape)
print("\nColumns:")
print(sleep_day.columns.tolist())
print("\nFirst 5 rows:")
print(sleep_day.head())

Shape: (413, 5)

Columns:
['Id', 'SleepDay', 'TotalSleepRecords', 'TotalMinutesAsleep', 'TotalTimeInBed']

First 5 rows:
           Id               SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0  1503960366  4/12/2016 12:00:00 AM                  1                 327   
1  1503960366  4/13/2016 12:00:00 AM                  2                 384   
2  1503960366  4/15/2016 12:00:00 AM                  1                 412   
3  1503960366  4/16/2016 12:00:00 AM                  2                 340   
4  1503960366  4/17/2016 12:00:00 AM                  1                 700   

   TotalTimeInBed  
0             346  
1             407  
2             442  
3             367  
4             712  


In [15]:
print("Missing values:")
print(sleep_day.isnull().sum())

print("\nDuplicate rows:")
print(sleep_day.duplicated().sum())

Missing values:
Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

Duplicate rows:
3


In [16]:
sleep_day = sleep_day.drop_duplicates().reset_index(drop=True)

print("Shape after removing duplicates:", sleep_day.shape)
print("Duplicate rows remaining:", sleep_day.duplicated().sum())

Shape after removing duplicates: (410, 5)
Duplicate rows remaining: 0


In [17]:
sleep_day["SleepDay"] = pd.to_datetime(sleep_day["SleepDay"])

print(sleep_day.dtypes)
print("\nDate range:")
print(sleep_day["SleepDay"].min())
print(sleep_day["SleepDay"].max())

Id                             int64
SleepDay              datetime64[ns]
TotalSleepRecords              int64
TotalMinutesAsleep             int64
TotalTimeInBed                 int64
dtype: object

Date range:
2016-04-12 00:00:00
2016-05-12 00:00:00


/tmp/ipykernel_58/745898383.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  sleep_day["SleepDay"] = pd.to_datetime(sleep_day["SleepDay"])


In [18]:
print("Unique users with sleep data:", sleep_day["Id"].nunique())

sleep_days_per_user = sleep_day.groupby("Id")["SleepDay"].count()

print("\nSleep days per user:")
print(sleep_days_per_user.describe())

Unique users with sleep data: 24

Sleep days per user:
count    24.000000
mean     17.083333
std      11.378533
min       1.000000
25%       4.750000
50%      20.500000
75%      27.250000
max      31.000000
Name: SleepDay, dtype: float64


In [19]:
sleep_duplicates = sleep_day.duplicated(
    subset=["Id", "SleepDay"]
).sum()

print("Duplicate user-date sleep records:", sleep_duplicates)

Duplicate user-date sleep records: 0


In [20]:
print("Average minutes asleep:",
      sleep_day["TotalMinutesAsleep"].mean())

print("Average hours asleep:",
      sleep_day["TotalMinutesAsleep"].mean() / 60)

print("\nAverage time in bed:",
      sleep_day["TotalTimeInBed"].mean() / 60)

Average minutes asleep: 419.17317073170733
Average hours asleep: 6.986219512195122

Average time in bed: 7.641382113821138


In [21]:
sleep_day["SleepEfficiency"] = (
    sleep_day["TotalMinutesAsleep"] /
    sleep_day["TotalTimeInBed"]
) * 100

print("Average sleep efficiency:",
      sleep_day["SleepEfficiency"].mean())

print("\nSleep efficiency summary:")
print(sleep_day["SleepEfficiency"].describe())

Average sleep efficiency: 91.64667080104189

Sleep efficiency summary:
count    410.000000
mean      91.646671
std        8.728386
min       49.836066
25%       91.181156
50%       94.263857
75%       96.064164
max      100.000000
Name: SleepEfficiency, dtype: float64


In [22]:
activity_sleep = pd.merge(
    daily_activity,
    sleep_day[
        [
            "Id",
            "SleepDay",
            "TotalMinutesAsleep",
            "TotalTimeInBed",
            "SleepEfficiency"
        ]
    ],
    left_on=["Id", "ActivityDate"],
    right_on=["Id", "SleepDay"],
    how="inner"
)

print("Merged dataset shape:", activity_sleep.shape)
print("Unique users:", activity_sleep["Id"].nunique())

print("\nFirst 5 rows:")
print(
    activity_sleep[
        [
            "Id",
            "ActivityDate",
            "TotalSteps",
            "Calories",
            "TotalMinutesAsleep",
            "SleepEfficiency"
        ]
    ].head()
)

Merged dataset shape: (410, 21)
Unique users: 24

First 5 rows:
           Id ActivityDate  TotalSteps  Calories  TotalMinutesAsleep  \
0  1503960366   2016-04-12       13162      1985                 327   
1  1503960366   2016-04-13       10735      1797                 384   
2  1503960366   2016-04-15        9762      1745                 412   
3  1503960366   2016-04-16       12669      1863                 340   
4  1503960366   2016-04-17        9705      1728                 700   

   SleepEfficiency  
0        94.508671  
1        94.348894  
2        93.212670  
3        92.643052  
4        98.314607  


In [23]:
sleep_activity_corr = activity_sleep[
    [
        "TotalMinutesAsleep",
        "TotalSteps",
        "Calories",
        "SedentaryMinutes"
    ]
].corr()

print(sleep_activity_corr)

                    TotalMinutesAsleep  TotalSteps  Calories  SedentaryMinutes
TotalMinutesAsleep            1.000000   -0.190344 -0.031699         -0.601073
TotalSteps                   -0.190344    1.000000  0.406301         -0.130036
Calories                     -0.031699    0.406301  1.000000          0.098656
SedentaryMinutes             -0.601073   -0.130036  0.098656          1.000000


In [24]:
activity_sleep["DayOfWeek"] = activity_sleep["ActivityDate"].dt.day_name()

sleep_by_day = activity_sleep.groupby("DayOfWeek")["TotalMinutesAsleep"].mean()

day_order = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday"
]

sleep_by_day = sleep_by_day.reindex(day_order)

print(sleep_by_day)

DayOfWeek
Monday       419.500000
Tuesday      404.538462
Wednesday    434.681818
Thursday     401.296875
Friday       405.421053
Saturday     419.070175
Sunday       452.745455
Name: TotalMinutesAsleep, dtype: float64


In [25]:
sleep_records_by_day = activity_sleep.groupby(
    "DayOfWeek"
)["TotalMinutesAsleep"].count()

sleep_records_by_day = sleep_records_by_day.reindex(day_order)

print(sleep_records_by_day)

DayOfWeek
Monday       46
Tuesday      65
Wednesday    66
Thursday     64
Friday       57
Saturday     57
Sunday       55
Name: TotalMinutesAsleep, dtype: int64


In [26]:
user_avg_steps = daily_activity.groupby("Id")["TotalSteps"].mean()

print(user_avg_steps.describe())

count       35.000000
mean      7078.943021
std       3842.094305
min        773.625000
25%       4472.042125
50%       7363.000000
75%       8671.869048
max      16759.428571
Name: TotalSteps, dtype: float64


In [27]:
user_activity = user_avg_steps.reset_index()
user_activity.columns = ["Id", "AverageDailySteps"]

def activity_level(steps):
    if steps < 5000:
        return "Low activity"
    elif steps < 10000:
        return "Moderate activity"
    else:
        return "High activity"

user_activity["ActivityLevel"] = user_activity[
    "AverageDailySteps"
].apply(activity_level)

print(user_activity["ActivityLevel"].value_counts())

print("\nPercentage of users:")
print(
    user_activity["ActivityLevel"]
    .value_counts(normalize=True)
    .mul(100)
    .round(1)
)

ActivityLevel
Moderate activity    17
Low activity         11
High activity         7
Name: count, dtype: int64

Percentage of users:
ActivityLevel
Moderate activity    48.6
Low activity         31.4
High activity        20.0
Name: proportion, dtype: float64


In [28]:
user_summary = daily_activity.groupby("Id").agg(
    AverageDailySteps=("TotalSteps", "mean"),
    DaysRecorded=("ActivityDate", "count")
).reset_index()

user_summary["ActivityLevel"] = user_summary[
    "AverageDailySteps"
].apply(activity_level)

print(
    user_summary.groupby("ActivityLevel")["DaysRecorded"]
    .agg(["count", "mean", "median", "min", "max"])
)

                   count       mean  median  min  max
ActivityLevel                                        
High activity          7  42.285714    42.0   37   49
Low activity          11  36.272727    40.0    8   62
Moderate activity     17  39.882353    42.0   30   49


## 4. Key Findings

The analysis identified five main findings:

1. **Weekly activity patterns:** Users averaged approximately 7,377 steps per day. Saturday had the highest average daily steps at 7,752, while Sunday had the lowest at 6,607.

2. **Activity levels:** Based on the project-defined activity segments, 48.6% of users were moderately active, 31.4% were low activity, and 20.0% were highly active. Overall, 80% of users were in the low or moderate activity groups.

3. **Activity intensity:** Light activity dominated recorded active time, averaging 188.1 minutes per day compared with 19.9 very-active minutes and 13.6 fairly-active minutes.

4. **Steps and calories:** Daily steps had a moderate positive correlation with calories burned (r = 0.58), indicating that higher-step days generally tended to be associated with greater calorie expenditure.

5. **Sleep patterns:** Among the 24 users with available sleep data, average sleep duration was approximately 6.99 hours per night. Sunday had the highest average sleep duration at 7.55 hours, while Thursday had the lowest at 6.69 hours.

In [29]:
steps_excel = (
    daily_activity.groupby("DayOfWeek")["TotalSteps"]
    .mean()
    .reindex(day_order)
    .reset_index()
)

steps_excel.columns = ["DayOfWeek", "AverageSteps"]

steps_excel["AverageSteps"] = (
    steps_excel["AverageSteps"].round(0)
)

print(steps_excel)

   DayOfWeek  AverageSteps
0     Monday        7541.0
1    Tuesday        7719.0
2  Wednesday        7548.0
3   Thursday        7268.0
4     Friday        7188.0
5   Saturday        7752.0
6     Sunday        6607.0


In [30]:
steps_excel.to_csv(
    "/kaggle/working/average_steps_by_day.csv",
    index=False
)

print("File exported successfully.")

File exported successfully.


In [31]:
activity_level_excel = (
    user_activity["ActivityLevel"]
    .value_counts()
    .reindex([
        "Low activity",
        "Moderate activity",
        "High activity"
    ])
    .reset_index()
)

activity_level_excel.columns = ["ActivityLevel", "NumberOfUsers"]

activity_level_excel["Percentage"] = (
    activity_level_excel["NumberOfUsers"]
    / activity_level_excel["NumberOfUsers"].sum()
    * 100
).round(1)

print(activity_level_excel)

       ActivityLevel  NumberOfUsers  Percentage
0       Low activity             11        31.4
1  Moderate activity             17        48.6
2      High activity              7        20.0


In [32]:
activity_level_excel.to_csv(
    "/kaggle/working/activity_level_distribution.csv",
    index=False
)

print("File exported successfully.")

File exported successfully.


In [33]:
activity_minutes_excel = pd.DataFrame({
    "ActivityLevel": [
        "Very Active",
        "Fairly Active",
        "Lightly Active"
    ],
    "AverageMinutes": [
        daily_activity["VeryActiveMinutes"].mean(),
        daily_activity["FairlyActiveMinutes"].mean(),
        daily_activity["LightlyActiveMinutes"].mean()
    ]
})

activity_minutes_excel["AverageMinutes"] = (
    activity_minutes_excel["AverageMinutes"].round(1)
)

print(activity_minutes_excel)

    ActivityLevel  AverageMinutes
0     Very Active            19.9
1   Fairly Active            13.6
2  Lightly Active           188.1


In [34]:
activity_minutes_excel.to_csv(
    "/kaggle/working/activity_minutes_by_intensity.csv",
    index=False
)

print("File exported successfully.")

File exported successfully.


In [35]:
steps_calories_excel = daily_activity[
    ["TotalSteps", "Calories"]
].copy()

print(steps_calories_excel.head())
print("Number of records:", len(steps_calories_excel))

   TotalSteps  Calories
0       11004      1819
1       17609      2154
2       12736      1944
3       13231      1932
4       12041      1886
Number of records: 1373


In [36]:
steps_calories_excel.to_csv(
    "/kaggle/working/steps_vs_calories.csv",
    index=False
)

print("File exported successfully.")

File exported successfully.


In [38]:
sleep_excel = (
    activity_sleep.groupby("DayOfWeek")["TotalMinutesAsleep"]
    .mean()
    .reindex(day_order)
    .reset_index()
)

sleep_excel["AverageSleepHours"] = (
    sleep_excel["TotalMinutesAsleep"] / 60
).round(2)

sleep_excel = sleep_excel[
    ["DayOfWeek", "AverageSleepHours"]
]

print(sleep_excel)

   DayOfWeek  AverageSleepHours
0     Monday               6.99
1    Tuesday               6.74
2  Wednesday               7.24
3   Thursday               6.69
4     Friday               6.76
5   Saturday               6.98
6     Sunday               7.55


In [39]:
sleep_excel.to_csv(
    "/kaggle/working/average_sleep_by_day.csv",
    index=False
)

print("File exported successfully.")

File exported successfully.


## 5. Business Recommendations

Based on the analysis, three recommendations are proposed for Bellabeat:

### 1. Personalize Achievable Movement Goals
Bellabeat could provide personalized daily movement goals based on users' activity levels. Since 80% of users in this analysis were classified as low or moderately active and light activity represented the largest portion of recorded active time, the app could encourage walking, short movement breaks, and gradual increases in daily activity.

### 2. Use Day-Specific Engagement
Bellabeat could use personalized reminders, challenges, and motivational messages based on users' activity patterns across the week. Rather than applying the same strategy to every user, engagement could adapt to each user's individual behavior.

### 3. Integrate Activity and Sleep Insights
Bellabeat could combine activity and sleep information to provide personalized weekly wellness summaries. This could help users better understand their movement and sleep patterns together.

## Conclusion

This analysis identified meaningful patterns in activity levels, activity intensity, calorie expenditure, and sleep behavior that could help inform Bellabeat's marketing and user-engagement strategy.

However, the findings are exploratory because the dataset represents a relatively small sample of Fitbit users rather than Bellabeat customers and was collected in 2016. Bellabeat should validate these patterns using larger, more recent data from its own users before making major marketing decisions.